hal/network/ncp_client/quectel/quectel_ncp_client.cpp

```diff
@@ -1371,12 +1400,50 @@ int QuectelNcpClient::enterDataMode() {
     const char connectResponse[] = "CONNECT";
 
     char buf[64] = {};
-    auto resp = dataParser_.sendCommand(1000, "ATO");
+    auto resp = dataParser_.sendCommand(2000, "ATO");
     if (resp.hasNextLine()) {
         CHECK(resp.readLine(buf, sizeof(buf)));
         if (!strncmp(buf, connectResponse, sizeof(connectResponse) - 1)) {
-            // We've already switched into data mode
-            ok = true;
+            // IMPORTANT: ATO does not work all the time and we resume into
+            // some broken state where the modem does not reply to our PPP ConfReqs.
+            // What's worse is that ATH doesn't seem to work either, so we cannot
+            // disconnect an ongoing PPP session. Toggling DTR doesn't work either.
+            // As a workaround send a PPP LCP Echo Request, wait until we get a reply
+            // or at least something similar to it.
+            // If we don't get a reply, consider we are in a broken state. To get
+            // out of it we are going to close and re-open the data muxer channel.
+            dataParser_.reset();
+            skipAll(muxerDataStream_.get());
+            int flagsSeen = 0;
+            for (int i = 0; i < PPP_ECHO_REQUEST_ATTEMPTS; i++) {
+                // I wish this was done in PPP layer
+                const char lcpEchoRequest[] = "~\xff\x03\xc0!\t\x00\x00\x08\x00\x00\x00\x00\xbbn~";
+                muxerDataStream_->write(lcpEchoRequest, sizeof(lcpEchoRequest) - 1);
+                const int r = muxerDataStream_->waitEvent(Stream::READABLE, 1000);
+                if (r == Stream::READABLE) {
+                    // Do a quick check that we've seen something similar to a PPP frame: ~<...>~
+                    const auto size = CHECK(muxerDataStream_->read(buf, sizeof(buf)));
+                    for (int i = 0; i < size; i++) {
+                        if (buf[i] == lcpEchoRequest[0]) {
+                            flagsSeen++;
+                        }
+                    }
+                    if (flagsSeen >= 2) {
+                        ok = true;
+                        break;
+                    }
+                } else if (r != SYSTEM_ERROR_TIMEOUT) {
+                    return r;
```

Comment:
So, just to clarify, we don't want to retry the echo request if we haven't received a reply for it within 1s?

Reply:
Correct. Or don't receive something similar to a PPP frame: 0x7e ...... 0x7e

Reply:
Ah, I misread it as `== SYSTEM_ERROR_TIMEOUT`, sorry.